common/src/main/java/org/apache/drill/common/scanner/persistence/ScanResult.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.common.scanner.persistence;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.apache.drill.common.exceptions.DrillRuntimeException;
+import org.apache.drill.common.scanner.ClassPathScanner;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+/**
+ * The root doc of the scan result
+ */
+public final class ScanResult {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ScanResult.class);
+  private final List<String> prescannedPackages;
+  private final List<String> scannedPackages;
+  private final List<FunctionDescriptor> functions;
+  private final List<ParentClassDescriptor> implementations;
+  private final Map<String, ParentClassDescriptor> parentClassByName;
+
+  @JsonCreator public ScanResult(
+      @JsonProperty("prescannedPackages") List<String> prescannedPackages,
+      @JsonProperty("scannedPackages") List<String> scannedPackages,
+      @JsonProperty("functions") List<FunctionDescriptor> functions,
```

Comment:
Let's simplify. We can find functions by their interface class, not their annotation. This simplifies scanning and what we hold onto. The org.apache.drill.exec.expr.DrillFunc is the marking interface.